From 60d4712d0cb4202df33ae6baf5c0010e70dab6f9 Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Mon, 19 Sep 2022 01:07:09 +0530
Subject: [PATCH] PIHooks: Spoof device as Pixel for some Google apps

Unlocks features like unlimited google photos backup,
new weather widgets introduced in Android 12, nicer
looking google feed (?), Pixel 6 wallpapers etc.

Move gphotos feature blacklist here while we're at it.

Change-Id: I21a2c676db6c5ccf27f8ea7ad27cbe6e0dcd6c25
Signed-off-by: penglezos <panagiotisegl@gmail.com>
---
 .../app/ApplicationPackageManager.java        |  4 +-
 .../internal/util/PropImitationHooks.java     | 55 +++++++++++++++++++
 core/res/res/values/config.xml                |  4 ++
 core/res/res/values/symbols.xml               |  3 +
 4 files changed, 65 insertions(+), 1 deletion(-)

diff --git a/core/java/android/app/ApplicationPackageManager.java b/core/java/android/app/ApplicationPackageManager.java
index 475e0f4cb05b..e55dff038a61 100644
--- a/core/java/android/app/ApplicationPackageManager.java
+++ b/core/java/android/app/ApplicationPackageManager.java
@@ -120,6 +120,7 @@ import com.android.internal.annotations.GuardedBy;
 import com.android.internal.annotations.Immutable;
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.os.SomeArgs;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.UserIcons;
 
 import com.nvidia.NvAppProfileService;
@@ -820,7 +821,8 @@ public class ApplicationPackageManager extends PackageManager {
 
     @Override
     public boolean hasSystemFeature(String name, int version) {
-        return mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version));
+        return PropImitationHooks.hasSystemFeature(name,
+            mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version)));
     }
 
     /** @hide */
diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index 69ca0f1fd4b9..fdb251a35589 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -25,6 +25,8 @@ import com.android.internal.R;
 
 import java.lang.reflect.Field;
 import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
 
 public class PropImitationHooks {
 
@@ -46,9 +48,45 @@ public class PropImitationHooks {
     private static final String PACKAGE_GMS = "com.google.android.gms";
     private static final String PROCESS_GMS_UNSTABLE = PACKAGE_GMS + ".unstable";
 
+    private static final String PACKAGE_GPHOTOS = "com.google.android.apps.photos";
+    private static final Map<String, Object> sP1Props = new HashMap<>();
+    static {
+        sP1Props.put("BRAND", "google");
+        sP1Props.put("MANUFACTURER", "Google");
+        sP1Props.put("DEVICE", "marlin");
+        sP1Props.put("PRODUCT", "marlin");
+        sP1Props.put("MODEL", "Pixel XL");
+        sP1Props.put("FINGERPRINT", "google/marlin/marlin:10/QP1A.191005.007.A3/5972272:user/release-keys");
+    }
+    private static final String[] sFeaturesBlacklist = {
+        "PIXEL_2017_PRELOAD",
+        "PIXEL_2018_PRELOAD",
+        "PIXEL_2019_MIDYEAR_PRELOAD",
+        "PIXEL_2019_PRELOAD",
+        "PIXEL_2020_EXPERIENCE",
+        "PIXEL_2020_MIDYEAR_EXPERIENCE",
+        "PIXEL_2021_EXPERIENCE",
+        "PIXEL_2021_MIDYEAR_EXPERIENCE"
+    };
+
+    private static final String PACKAGE_WALLPAPERS = "com.google.android.apps.wallpaper";
+    private static final Map<String, Object> sP6Props = new HashMap<>();
+    static {
+        sP6Props.put("BRAND", "google");
+        sP6Props.put("MANUFACTURER", "Google");
+        sP6Props.put("DEVICE", "raven");
+        sP6Props.put("PRODUCT", "raven");
+        sP6Props.put("MODEL", "Pixel 6 Pro");
+        sP6Props.put("FINGERPRINT", "google/raven/raven:13/TP1A.220624.021/8877034:user/release-keys");
+    }
+
+    private static final boolean sSpoofGapps =
+            Resources.getSystem().getBoolean(R.bool.config_spoofGoogleApps);
+
     private static volatile boolean sIsGms = false;
     private static volatile boolean sIsFinsky = false;
     private static volatile boolean sIsVelvet = false;
+    private static volatile boolean sIsPhotos = false;
 
     public static void setProps(Application app) {
         final String packageName = app.getPackageName();
@@ -61,6 +99,7 @@ public class PropImitationHooks {
         sIsGms = packageName.equals(PACKAGE_GMS) && processName.equals(PROCESS_GMS_UNSTABLE);
         sIsFinsky = packageName.equals(PACKAGE_FINSKY);
         sIsVelvet = packageName.equals(PACKAGE_VELVET);
+        sIsPhotos = sSpoofGapps && packageName.equals(PACKAGE_GPHOTOS);
 
         /* Set Certified Fingerprint for GMSCore or Finsky
          * Set Certified Fingerprint and Model for Velvet
@@ -77,6 +116,13 @@ public class PropImitationHooks {
         } else if (!sStockFp.isEmpty() && packageName.equals(PACKAGE_ARCORE)) {
             dlog("Setting stock fingerprint for: " + packageName);
             setPropValue("FINGERPRINT", sStockFp);
+        } else if (sIsPhotos) {
+            dlog("Spoofing Pixel XL for Google Photos");
+            sP1Props.forEach((k, v) -> setPropValue(k, v));
+        } else if (sSpoofGapps && (packageName.equals(PACKAGE_VELVET)
+                || packageName.equals(PACKAGE_WALLPAPERS))) {
+            dlog("Spoofing Pixel 6 Pro for: " + packageName);
+            sP6Props.forEach((k, v) -> setPropValue(k, v));
         }
     }
 
@@ -105,6 +151,15 @@ public class PropImitationHooks {
         }
     }
 
+    public static boolean hasSystemFeature(String name, boolean def) {
+        if (sIsPhotos && def &&
+                Arrays.stream(sFeaturesBlacklist).anyMatch(name::contains)) {
+            dlog("Blocked system feature " + name + " for Google Photos");
+            return false;
+        }
+        return def;
+    }
+
     public static void dlog(String msg) {
       if (DEBUG) Log.d(TAG, msg);
     }
diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index a7664d7e783e..e1cc2a2a8626 100644
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -5944,4 +5944,8 @@
 
     <!-- Fingerprint from stock ROM -->
     <string name="config_stockFingerprint" translatable="false"></string>
+     
+     <!-- Whether to spoof device as pixel for certain google apps, to
+     unlock pixel-exclusive features -->
+    <bool name="config_spoofGoogleApps">true</bool>
 </resources>
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
index 53c64e9393aa..3f407f44b23a 100644
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -4864,4 +4864,7 @@
 
   <!-- Fingerprint from stock ROM -->
   <java-symbol type="string" name="config_stockFingerprint" />
+
+  <!-- Spoof Google Apps -->
+  <java-symbol type="bool" name="config_spoofGoogleApps" />
 </resources>
-- 
2.38.2

